packages/@n8n/task-runner-python/src/task_analyzer.py

```diff
@@ -0,0 +1,205 @@
+import ast
+import hashlib
+import sys
+from typing import Set, Tuple
+from collections import OrderedDict
+
+from src.errors import SecurityViolationError
+from src.constants import MAX_VALIDATION_CACHE_SIZE
+
+CacheKey = Tuple[str, Tuple]  # (code_hash, allowlists_tuple)
+CachedViolations = list[str]
+ValidationCache = OrderedDict[CacheKey, CachedViolations]
+
+ERROR_RELATIVE_IMPORT = "Relative imports are disallowed."
+ERROR_DYNAMIC_IMPORT = "Dynamic imports using `__import__()` with variables are disallowed, because they can bypass allowlist validation at runtime."
+ERROR_DYNAMIC_IMPORTLIB = "Dynamic imports using `importlib` with variables are disallowed, because they can bypass allowlist validation at runtime."
+ERROR_STDLIB_DISALLOWED = "Import of standard library module '{module}' is disallowed. Allowed stdlib modules: {allowed}"
+ERROR_EXTERNAL_DISALLOWED = "Import of external package '{module}' is disallowed. Allowed external packages: {allowed}"
+ERROR_SECURITY_VIOLATIONS = "Security violations detected:\n{violations}"
+
+
+class ImportValidator(ast.NodeVisitor):
+    def __init__(self, stdlib_allow: Set[str], external_allow: Set[str]):
+        self.checked_modules: Set[str] = set()
+        self.violations: list[str] = []
+
+        self.stdlib_allow = stdlib_allow
+        self.external_allow = external_allow
+        self._stdlib_allowed_str = self._format_allowed(stdlib_allow)
+        self._external_allowed_str = self._format_allowed(external_allow)
+        self._stdlib_allow_all = "*" in stdlib_allow
+        self._external_allow_all = "*" in external_allow
+
+    # ========== Visitors ==========
```

Comment:
Maybe we can make this and other comment "tags" regions? I.e. `#region ======== Visitors =========` with an `#endregion` so they are foldable in VSCode

Reply:
If it's okay I'll skip this one. `region` doesn't work in nvim and the intent is mostly keeping visual hierarchy as you're scrolling, rather than folding bundles of methods, which is likely very rarely used.